services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -0,0 +1,256 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.internal.handlers.EndpointAddressInterceptor;
+import software.amazon.awssdk.services.s3.model.GetUrlRequest;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Utilities for working with Amazon S3 objects. An instance of this class can be created by:
+ * 1) Directly using the {@link #builder()} method. You have to manually specify the configuration params like region,
+ * s3Configuration on the builder.
+ * 2) Using the low-level client {@link S3Client#utilities()} method. This is recommended as SDK will use the same
+ * configuration from the {@link S3Client} object to create the {@link S3Utilities} object.
+ *
+ * Note: This class should not have methods that make network calls.
+ */
+@Immutable
+@SdkPublicApi
+public final class S3Utilities {
+
+    private final Region region;
+
+    private final S3Configuration s3Configuration;
+
+    /**
+     * SDK currently validates that region is present while constructing {@link S3Utilities} object.
+     * This can be relaxed in future when more methods are added that doesn't use region.
+     */
+    public S3Utilities(Builder builder) {
+        this.region = Validate.paramNotNull(builder.region, "Region");
+        this.s3Configuration = builder.s3Configuration;
+    }
+
+    /**
+     * Creates a builder for {@link S3Utilities}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    // Used by low-level client
+    @SdkInternalApi
+    static S3Utilities create(SdkClientConfiguration clientConfiguration) {
+        return S3Utilities.builder()
+                          .region(clientConfiguration.option(AwsClientOption.AWS_REGION))
+                          .s3Configuration((S3Configuration) clientConfiguration.option(SdkClientOption.SERVICE_CONFIGURATION))
+                          .build();
+    }
+
+    /**
+     * Returns the URL for an object stored in Amazon S3.
+     *
+     * If the object identified by the given bucket and key has public read permissions,
+     * then this URL can be directly accessed to retrieve the object's data.
+     *
+     * <p>
+     *     If same configuration options are set on both #GetUrlRequest and #S3Utilities objects (for example: region),
+     *     the configuration set on the #GetUrlRequest takes precedence.
+     * </p>
+     *
+     * <p>
+     *     This is a convenience which creates an instance of the {@link GetUrlRequest.Builder} avoiding the need to
+     *     create one manually via {@link GetUrlRequest#builder()}
+     * </p>
+     *
+     * @param getUrlRequest A {@link Consumer} that will call methods on {@link GetUrlRequest.Builder} to create a request.
+     * @return A URL for an object stored in Amazon S3.
+     * @throws MalformedURLException Generated Url is malformed
+     */
+    public URL getUrl(Consumer<GetUrlRequest.Builder> getUrlRequest) throws MalformedURLException {
```

Comment:
If we are creating the URL, should we make sure `MalformedURLException` is not thrown by validating the input?

Reply:
It is thrown when converting URI to URL `httpRequest.getUri().toURL()`. There is nothing to avoid this unless returning URI as output instead of URL.

Reply:
In reality, this exception would not happen since we are the one that are creating the URL for them, right? I just feel we should not put `throws MalformatedURLException` on the signature because if I was the customer, I would not expect `MalformedURLException` to be thrown from this method, can we try catch it and wrap it with SdkClientException?

Reply:
I don't want to hide it in the  SdkClientException. If there is malformed url (for whatever reason), its good for customer to handle it.

Reply:
Well, we are not exactly hiding it and are just wrapping it with `SdkClientException` so that customers can use the same exception handling logic like other SDK operations.

Reply:
+1, we shouldn't throw checked exceptions.

Reply:
In addition, we decided that S3Utilities is not a client and so `SdkClientException` doesn't make sense.

Reply:
SdkException ;)?

Reply:
Alternatively, IllegalArgumentException or IllegalStateException. They're not passing a URL, so getting back an IllegalURLException seems very weird.

Reply:
Hmm, I'd imagine `SdkClientException` applies to all **SDK-side** exceptions compared with `SdkServiceException` and doesn't have to be only limited for the actual service client. We are throwing `SdkClientException` in some of our utils methods as well, technically, customers can use the utils by itself and not using an actual service client. 

https://github.com/aws/aws-sdk-java-v2/blob/d17d35f52bd197da33572e5b7854ed54750fd2dc/core/sdk-core/src/main/java/software/amazon/awssdk/core/util/json/JacksonUtils.java#L72

I'm also fine with `SdkException`.